addon/-private/object.ts

```diff
@@ -3,16 +3,88 @@ import { assert } from '@ember/debug';
 import { USE_PROXY } from './config';
 import { consume, dirty, SUPPORTS_PROXY } from './util';
 
+const setterPropertyPrefix = '__PROPERTY__';
+
+function defineHiddenProperty(obj: object, propertyName: string, value: any) {
+  Object.defineProperty(obj, propertyName, {
+    enumerable: false,
+    configurable: true,
+    writable: true,
+    value
+  })
+}
+
+function installSetter(obj: object, keyName: string, desc: PropertyDescriptor) {
+  const oldSet = desc.set || undefined;
+  if (typeof oldSet === 'function') {
+    desc.set = function(value: any) {
+      dirty(this, keyName);
+      return oldSet.call(this, value);
+    }
+  } else {
+    const key = `${setterPropertyPrefix}${keyName}`;
+    if (!(key in obj)) {
+      defineHiddenProperty(obj, key, desc.value);
+    }
+    desc.set = function(value: any) {
+      dirty(this, keyName);
+      (this as any)[key] = value;
+      return true;
+    }
+  }
+}
+
+function installGetter(obj: object, keyName: string, desc: PropertyDescriptor) {
+  const oldGet = desc.get || undefined;
+  if (typeof oldGet === 'function') {
+    desc.get = function() {
+      consume(this, keyName);
+      return oldGet.call(this);
+    }
+  } else {
+    const key = `${setterPropertyPrefix}${keyName}`;
+    if (!(key in obj)) {
+      defineHiddenProperty(obj, key, desc.value);
+    }
+    desc.get = function() {
+      consume(this, keyName);
+      return (this as any)[key];
+    }
+  }
+}
+
+function patchObjectDescriptor(obj: object, keyName: string, desc: PropertyDescriptor) {
+  if (desc.configurable === true) {
+    installSetter(obj, keyName, desc);
+    installGetter(obj, keyName, desc);
+    if ('value' in desc) {
+      delete desc.value;
+    }
+    if ('writable' in desc) {
+      delete desc.writable;
+    }
+    Object.defineProperty(obj, keyName, desc);
+  }
+}
+
 export default class TrackedObject extends Object {
-  constructor(obj: object) {
+  constructor(obj: object = {}) {
     super();
 
-    Object.assign(this, obj);
-
     assert(
       'You attempted to create a TrackedObject, but you have not enabled Proxy support or you are using a browser that does not support Proxy.',
-      SUPPORTS_PROXY && USE_PROXY
+      USE_PROXY ? SUPPORTS_PROXY : true
     );
+    
+    Object.assign(this, obj);
+
+    if (USE_PROXY === false) {
+      Object.keys(obj).forEach((keyName)=>{
+        const desc: PropertyDescriptor = Object.getOwnPropertyDescriptor(obj, keyName) as PropertyDescriptor;
+        patchObjectDescriptor(this, keyName, desc);
+      });
+      return this;
```

Comment:
I think it may make sense to add an `Object.seal(this);` here in DEBUG mode, so users don't accidentally try to add values to the object which won't be tracked, what do you think?